File: latex/main paper/python code/tailrecu.py
from random import random as U
import numpy as np
def X(t, a) -> np.array:
    q, A = np.array([1.0, 0.0]), np.array([[a, 0.0], [1.0, a]])
    sol, W = np.array([1.0, 0.0]), np.identity(2)
    while U() < t:
        W = W @ A if t < 1 else t * W @ A
        sol += W @ q
        t *= U()
    return sol
def sol(t, a, nsim): return sum(X(t, a) for _ in range(nsim))/nsim
print(f"x(1,1) = {sol(1,1,10**3)}")
# x(1,1) = [2.7198 2.7163]

File: latex/main paper/julia_code/tailrecu.jl
function X(t, a)
    (q = [1, 0]; A = [a 0; 1 a])
    (sol = q; W = [1.0 0.0; 0.0 1.0])
    while rand() < t
        W = W * A
        sol += W * q
        t *= rand()
    end
    sol
end